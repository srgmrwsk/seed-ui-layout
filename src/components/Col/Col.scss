.seed-col {
  flex-shrink: 0;
  width: 100%;
  max-width: 100%;
  min-width: 0;
  line-height: 1;

  @for $i from 1 through $grid-cols {
    &:first-of-type:nth-last-of-type(#{$i}n),
    &:first-of-type:nth-last-of-type(#{$i}n) ~ * {
      width: 100% / $i;
    }
  }

  @each $breakpoint in map-keys($grid-breakpoints) {
    @include bp-up($breakpoint) {
      &_#{$breakpoint} {
        &_fill {
          flex: 1 0 0;
          display: block;
        }

        &_auto {
          flex: 0 0 auto;
          display: block;
          width: auto;
        }

        &_none {
          display: none;
        }

        @for $i from 1 through $grid-cols {
          &_#{$i} {
            flex: 0 0 auto;
            display: block;
            width: $i / 12 * 100%;
          }
        }
      }
    }
  }
}
