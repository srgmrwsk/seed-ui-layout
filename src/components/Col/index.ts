export * from './types/ColSpan';
export * from './Col';
export { default as Col } from './Col';
