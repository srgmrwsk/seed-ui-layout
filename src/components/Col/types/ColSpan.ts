export type ColSpan = boolean | number | 'auto';
