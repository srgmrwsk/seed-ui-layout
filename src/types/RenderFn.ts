import React from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RenderFn = (component: React.ReactElement, ...args: []) => any;
