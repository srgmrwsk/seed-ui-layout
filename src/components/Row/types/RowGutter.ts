import { Sizing } from '../../../types/Sizing';

export type RowGutter = Sizing | [Sizing, Sizing];
