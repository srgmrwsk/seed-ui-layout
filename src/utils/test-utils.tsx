import React from 'react';
import { render } from '@testing-library/react';
import { renderToStaticMarkup } from 'react-dom/server';
import { configureAxe } from 'jest-axe';

import { RenderFn } from '../types/RenderFn';

export const axe = configureAxe({
  rules: {
    // disable landmark rules when testing isolated components
    region: { enabled: false },
  },
});

export function create(
  ...args: Parameters<RenderFn>
): ChildNode | HTMLCollection | null {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
}
export function renderToHtml(component: React.ReactElement): string {
  return renderToStaticMarkup(component);
}
