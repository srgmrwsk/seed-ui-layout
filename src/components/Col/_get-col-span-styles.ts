import { Maybe } from '../../types/Maybe';
import { getStylesBy } from '../../utils/get-styles-by';

import { ColSpan } from './types/ColSpan';

function getColSpan(value: Maybe<ColSpan>): Maybe<string> {
  return (
    (value === true && 'fill') ||
    ((value === false || value === 0) && 'none') ||
    (typeof value === 'number' && value.toString()) ||
    null
  );
}

export const getColSpanStyles = getStylesBy(getColSpan);
