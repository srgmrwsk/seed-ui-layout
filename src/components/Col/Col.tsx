import React from 'react';
import cx from 'classnames';

import { ResponsiveValue } from '../../types/ResponsiveValue';
import { Align } from '../../types/Align';
import { getStringStyles } from '../../utils/get-string-styles';
import { getOrderVariables } from '../../utils/get-order-variables';
import { getBooleanStyles } from '../../utils/get-boolean-styles';

import { getColSpanStyles } from './_get-col-span-styles';
import { ColSpan } from './types/ColSpan';

export interface ColProps
  extends Omit<React.AllHTMLAttributes<HTMLElement>, 'as'> {
  as?: React.ElementType;
  xs?: ColSpan;
  sm?: ColSpan;
  md?: ColSpan;
  lg?: ColSpan;
  xl?: ColSpan;
  xxl?: ColSpan;
  order?: ResponsiveValue<number>;
  align?: ResponsiveValue<Align>;
}

function Col(
  {
    as: As = 'div',
    xs,
    sm,
    md,
    lg,
    xl,
    xxl,
    align,
    order,
    className,
    style,
    children,
    ...elementProps
  }: ColProps,
  ref: React.Ref<HTMLElement>,
): JSX.Element {
  return (
    <As
      {...elementProps}
      ref={ref}
      className={cx(
        'seed-col',
        getColSpanStyles('seed-col_xs', xs),
        getColSpanStyles('seed-col_sm', sm),
        getColSpanStyles('seed-col_md', md),
        getColSpanStyles('seed-col_lg', lg),
        getColSpanStyles('seed-col_xl', xl),
        getColSpanStyles('seed-col_xxl', xxl),
        getStringStyles('_seed-align-self', align),
        getBooleanStyles('_seed-order', order),
        className,
      )}
      style={{
        ...getOrderVariables('seed-order', order),
        ...style,
      }}
    >
      {children}
    </As>
  );
}

export default React.forwardRef(Col);
