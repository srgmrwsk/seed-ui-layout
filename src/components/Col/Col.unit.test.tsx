import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { RenderFn } from '../../types/RenderFn';
import { axe, create } from '../../utils/test-utils';

import Col, { ColProps } from './Col';

describe('Col', () => {
  const setup = (renderFn: RenderFn, props?: ColProps) =>
    renderFn(<Col {...props} />);

  it('should render correctly with default props', () => {
    const actual = setup(create);
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with as prop', () => {
    const actual = setup(create, { as: 'article' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children prop', () => {
    const actual = setup(create, { children: <p>Element with text.</p> });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop xs
   */

  it('should render correctly with xs number prop', () => {
    const actual = setup(create, { xs: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xs true prop', () => {
    const actual = setup(create, { xs: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xs false prop', () => {
    const actual = setup(create, { xs: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xs auto prop', () => {
    const actual = setup(create, { xs: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop sm
   */

  it('should render correctly with sm number prop', () => {
    const actual = setup(create, { sm: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with sm true prop', () => {
    const actual = setup(create, { sm: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with sm false prop', () => {
    const actual = setup(create, { sm: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with sm auto prop', () => {
    const actual = setup(create, { sm: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop md
   */

  it('should render correctly with md number prop', () => {
    const actual = setup(create, { md: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with md true prop', () => {
    const actual = setup(create, { md: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with md false prop', () => {
    const actual = setup(create, { md: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with md auto prop', () => {
    const actual = setup(create, { md: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop lg
   */

  it('should render correctly with lg number prop', () => {
    const actual = setup(create, { lg: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with lg true prop', () => {
    const actual = setup(create, { lg: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with lg false prop', () => {
    const actual = setup(create, { lg: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with lg auto prop', () => {
    const actual = setup(create, { lg: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop xl
   */

  it('should render correctly with xl number prop', () => {
    const actual = setup(create, { xl: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xl true prop', () => {
    const actual = setup(create, { xl: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xl false prop', () => {
    const actual = setup(create, { xl: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xl auto prop', () => {
    const actual = setup(create, { xl: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop xxl
   */

  it('should render correctly with xxl number prop', () => {
    const actual = setup(create, { xxl: 3 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xxl true prop', () => {
    const actual = setup(create, { xxl: true });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xxl false prop', () => {
    const actual = setup(create, { xxl: false });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with xxl auto prop', () => {
    const actual = setup(create, { xxl: 'auto' });
    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop align
   */

  it('should render correctly with align prop', () => {
    const actual = setup(create, { align: 'center' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with align responsive prop', () => {
    const actual = setup(create, {
      align: {
        xs: 'baseline',
        sm: 'center',
        md: 'end',
        lg: 'start',
        xl: 'stretch',
        xxl: 'baseline',
      },
    });

    expect(actual).toMatchSnapshot();
  });

  /**
   *  @prop order
   */

  it('should render correctly with order prop', () => {
    const actual = setup(create, { order: 2 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with order responsive prop', () => {
    const actual = setup(create, {
      order: {
        xs: 1,
        sm: 2,
        md: 3,
        lg: 4,
        xl: 5,
        xxl: 6,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = setup(renderToStaticMarkup);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
