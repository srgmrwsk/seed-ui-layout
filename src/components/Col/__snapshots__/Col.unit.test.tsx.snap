// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col should render correctly with align prop 1`] = `
<div
  class="seed-col _seed-align-self_center"
/>
`;

exports[`Col should render correctly with align responsive prop 1`] = `
<div
  class="seed-col _seed-align-self_baseline _seed-align-self-sm_center _seed-align-self-md_end _seed-align-self-lg_start _seed-align-self-xl_stretch _seed-align-self-xxl_baseline"
/>
`;

exports[`Col should render correctly with as prop 1`] = `
<article
  class="seed-col"
/>
`;

exports[`Col should render correctly with children prop 1`] = `
<div
  class="seed-col"
>
  <p>
    Element with text.
  </p>
</div>
`;

exports[`Col should render correctly with default props 1`] = `
<div
  class="seed-col"
/>
`;

exports[`Col should render correctly with lg auto prop 1`] = `
<div
  class="seed-col seed-col_lg_auto"
/>
`;

exports[`Col should render correctly with lg false prop 1`] = `
<div
  class="seed-col seed-col_lg_none"
/>
`;

exports[`Col should render correctly with lg number prop 1`] = `
<div
  class="seed-col seed-col_lg_3"
/>
`;

exports[`Col should render correctly with lg true prop 1`] = `
<div
  class="seed-col seed-col_lg_fill"
/>
`;

exports[`Col should render correctly with md auto prop 1`] = `
<div
  class="seed-col seed-col_md_auto"
/>
`;

exports[`Col should render correctly with md false prop 1`] = `
<div
  class="seed-col seed-col_md_none"
/>
`;

exports[`Col should render correctly with md number prop 1`] = `
<div
  class="seed-col seed-col_md_3"
/>
`;

exports[`Col should render correctly with md true prop 1`] = `
<div
  class="seed-col seed-col_md_fill"
/>
`;

exports[`Col should render correctly with order prop 1`] = `
<div
  class="seed-col _seed-order"
  style="--seed-order: 2;"
/>
`;

exports[`Col should render correctly with order responsive prop 1`] = `
<div
  class="seed-col _seed-order _seed-order-sm _seed-order-md _seed-order-lg _seed-order-xl _seed-order-xxl"
  style="--seed-order: 1; --seed-order-sm: 2; --seed-order-md: 3; --seed-order-lg: 4; --seed-order-xl: 5; --seed-order-xxl: 6;"
/>
`;

exports[`Col should render correctly with sm auto prop 1`] = `
<div
  class="seed-col seed-col_sm_auto"
/>
`;

exports[`Col should render correctly with sm false prop 1`] = `
<div
  class="seed-col seed-col_sm_none"
/>
`;

exports[`Col should render correctly with sm number prop 1`] = `
<div
  class="seed-col seed-col_sm_3"
/>
`;

exports[`Col should render correctly with sm true prop 1`] = `
<div
  class="seed-col seed-col_sm_fill"
/>
`;

exports[`Col should render correctly with xl auto prop 1`] = `
<div
  class="seed-col seed-col_xl_auto"
/>
`;

exports[`Col should render correctly with xl false prop 1`] = `
<div
  class="seed-col seed-col_xl_none"
/>
`;

exports[`Col should render correctly with xl number prop 1`] = `
<div
  class="seed-col seed-col_xl_3"
/>
`;

exports[`Col should render correctly with xl true prop 1`] = `
<div
  class="seed-col seed-col_xl_fill"
/>
`;

exports[`Col should render correctly with xs auto prop 1`] = `
<div
  class="seed-col seed-col_xs_auto"
/>
`;

exports[`Col should render correctly with xs false prop 1`] = `
<div
  class="seed-col seed-col_xs_none"
/>
`;

exports[`Col should render correctly with xs number prop 1`] = `
<div
  class="seed-col seed-col_xs_3"
/>
`;

exports[`Col should render correctly with xs true prop 1`] = `
<div
  class="seed-col seed-col_xs_fill"
/>
`;

exports[`Col should render correctly with xxl auto prop 1`] = `
<div
  class="seed-col seed-col_xxl_auto"
/>
`;

exports[`Col should render correctly with xxl false prop 1`] = `
<div
  class="seed-col seed-col_xxl_none"
/>
`;

exports[`Col should render correctly with xxl number prop 1`] = `
<div
  class="seed-col seed-col_xxl_3"
/>
`;

exports[`Col should render correctly with xxl true prop 1`] = `
<div
  class="seed-col seed-col_xxl_fill"
/>
`;
